src/herder/Upgrades.cpp

```diff
@@ -318,4 +328,237 @@ Upgrades::timeForUpgrade(uint64_t time) const
+        "(ledgerseq, upgradeindex,  upgrade,  changes) VALUES "
+        "(:seq,      :upgradeindex, :upgrade, :changes)");
+
+    auto& st = prep.statement();
+    st.exchange(soci::use(ledgerSeq));
+    st.exchange(soci::use(index));
+    st.exchange(soci::use(upgradeContent64));
+    st.exchange(soci::use(upgradeChanges64));
+    st.define_and_bind();
+    {
+        auto timer = db.getInsertTimer("upgradehistory");
+        st.execute(true);
+    }
+
+    if (st.get_affected_rows() != 1)
+    {
+        throw std::runtime_error("Could not update data in SQL");
+    }
+}
+
+static void
+addLiabilities(std::map<Asset, std::unique_ptr<int64_t>>& liabilities,
+               AccountID const& accountID, Asset const& asset, int64_t delta)
+{
+    auto iter =
+        liabilities.insert(std::make_pair(asset, std::make_unique<int64_t>(0)))
+            .first;
+    if (asset.type() != ASSET_TYPE_NATIVE && accountID == getIssuer(asset))
+    {
+        return;
+    }
+    if (iter->second)
+    {
+        if (!stellar::addBalance(*iter->second, delta))
+        {
+            iter->second.reset();
+        }
+    }
+}
+
+static int64_t
+getAvailableBalance(AccountID const& accountID, Asset const& asset,
+                    int64_t balanceAboveReserve, LedgerDelta& ld, Database& db)
+{
+    if (asset.type() == ASSET_TYPE_NATIVE)
+    {
+        return balanceAboveReserve;
+    }
+
+    if (accountID == getIssuer(asset))
+    {
+        return INT64_MAX;
+    }
+    else
+    {
+        auto trust = TrustFrame::loadTrustLine(accountID, asset, db, ld);
+        assert(trust);
+        return trust->getBalance();
+    }
+}
+
+static int64_t
+getAvailableLimit(AccountID const& accountID, Asset const& asset,
+                  int64_t balanceAboveReserve, LedgerDelta& ld, Database& db)
+{
+    if (asset.type() == ASSET_TYPE_NATIVE)
+    {
+        return INT64_MAX - balanceAboveReserve;
+    }
+
+    if (accountID == getIssuer(asset))
+    {
+        return INT64_MAX;
+    }
+    else
+    {
+        auto trust = TrustFrame::loadTrustLine(accountID, asset, db, ld);
+        assert(trust);
+        return trust->getTrustLine().limit - trust->getBalance();
+    }
+}
+
+static bool
+shouldDeleteOffer(AccountID const& accountID, Asset const& asset,
+                  int64_t balanceAboveReserve,
+                  std::map<Asset, std::unique_ptr<int64_t>> const& liabilities,
+                  std::function<int64_t(AccountID const&, Asset const&, int64_t,
+                                        LedgerDelta&, Database&)>
+                      getCap,
+                  LedgerDelta& ld, Database& db)
+{
+    auto iter = liabilities.find(asset);
+    assert(iter != liabilities.end());
+    return iter->second ? *iter->second > getCap(accountID, asset,
+                                                 balanceAboveReserve, ld, db)
+                        : true;
+}
+
+static void
+prepareLiabilities(LedgerManager& ledgerManager, LedgerDelta& ld)
```

Comment:
I'm not sure if I understand this method correctly, please let me know if this is correct:

For each account, remove all offers that are using assets, for which the liabilities are too big. So it does not keep any offers from that set, even if keeping only some of them would make all the liabilities proper?